src/fips_check.c

```diff
@@ -0,0 +1,224 @@
+/*
+ * This program is released under the Common Public License V1.0
+ *
+ * You should have received a copy of Common Public License V1.0 along with
+ * with this program.
+ *
+ * Copyright IBM Corp. 2020
+ */
+
+
+#ifdef ICA_FIPS
+
+#include <errno.h>
+#include <openssl/crypto.h>
+#include <openssl/hmac.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <syslog.h>
+#include <dlfcn.h>
+
+
+#ifndef PATH_MAX
+#define PATH_MAX 4096
+#endif
+
+#define HMAC_PREFIX "."
+#ifndef HMAC_SUFFIX
+#define HMAC_SUFFIX ".hmac"
+#endif
+#define READ_BUFFER_LENGTH 16384
+
+static const char hmackey[] = "00000000000000000000000000000000";
+
+static int get_library_path(const char *libname, const char *symbolname,
+							char *path, size_t pathlen)
+{
+	Dl_info info;
+	void *dl, *sym;
+	int rc = -1;
+
+	dl = dlopen(libname, RTLD_LAZY);
+	if (dl == NULL)
+		goto done;
+
+	sym = dlsym(dl, symbolname);
+	if (sym != NULL && dladdr(sym, &info)) {
+		if (strlen(info.dli_fname) <= pathlen)
+			strcpy(path, info.dli_fname);
+		else
+			goto done;
+	}
+
+	dlclose(dl);
+	rc = 0;
+
+done:
+
+	return rc;
+}
+
+static char *make_hmac_path(const char *origpath)
+{
+	char *path, *p;
+	const char *fn;
+
+	path = malloc(sizeof(HMAC_PREFIX) + sizeof(HMAC_SUFFIX) + strlen(origpath));
+	if (path == NULL)
+		return NULL;
+
+	fn = strrchr(origpath, '/');
+	if (fn == NULL) {
+		fn = origpath;
+	} else {
+		++fn;
+	}
+
+	strncpy(path, origpath, fn - origpath);
+	p = path + (fn - origpath);
```

Comment:
why not use strcat to concat the strings and get rid of tracking the string's final '\0' byte with p ?

Reply:
Done.